pkg/koordlet/resmanager/resctrl_reconcile.go

```diff
@@ -0,0 +1,535 @@
+	nodeCPUInfo, err := r.resManager.metricCache.GetNodeCPUInfo(&metriccache.QueryParam{})
+	if err != nil {
+		klog.Warningf("failed to get nodeCPUInfo, err: %v", err)
+		return
+	}
+	if nodeCPUInfo == nil {
+		klog.Warning("failed to get nodeCPUInfo, the value is nil")
+		return
+	}
+	cbmStr := nodeCPUInfo.BasicInfo.CatL3CbmMask
+	if len(cbmStr) <= 0 {
+		klog.Warning("failed to get cat l3 cbm, cbm is empty")
+		return
+	}
+	cbmValue, err := strconv.ParseUint(cbmStr, 16, 32)
+	if err != nil {
+		klog.Warningf("failed to parse cat l3 cbm %s, err: %v", cbmStr, err)
+		return
+	}
+	cbm := uint(cbmValue)
+
+	// get the number of l3 caches; it is larger than 0
+	l3Num := int(nodeCPUInfo.TotalInfo.NumberL3s)
+	if l3Num <= 0 {
+		klog.Warningf("failed to get the number of l3 caches, invalid value %v", l3Num)
+		return
+	}
+
+	// calculate and apply l3 cat policy for each group
+	for _, group := range resctrlGroupList {
+		resQOSStrategy := getResourceQoSForResctrlGroup(qosStrategy, group)
+		err = r.calculateAndApplyCatL3PolicyForGroup(group, cbm, l3Num, resQOSStrategy)
+		if err != nil {
+			klog.Warningf("failed to apply l3 cat policy for group %v, err: %v", group, err)
+		}
+		err = r.calculateAndApplyCatMbPolicyForGroup(group, l3Num, resQOSStrategy)
+		if err != nil {
+			klog.Warningf("failed to apply cat MB policy for group %v, err: %v", group, err)
+		}
+	}
+}
+
+func (r *ResctrlReconcile) reconcileResctrlGroups(qosStrategy *slov1alpha1.ResourceQoSStrategy) {
+	// 1. retrieve task ids for each slo by reading cgroup task file of every pod container
+	// 2. add the related task ids in resctrl groups
+
+	// NOTE: pid_max can be found in `/proc/sys/kernel/pid_max` on linux.
+	// the maximum pid on 32-bit/64-bit platforms is always less than 4194304, so the int type is bigger enough.
+	// here we only append the task ids which only appear in cgroup but not in resctrl to reduce resctrl writes
+	var err error
+
+	curTaskMaps := map[string]map[int]struct{}{}
+	for _, group := range resctrlGroupList {
+		curTaskMaps[group], err = system.ReadResctrlTasksMap(group)
+		if err != nil {
+			klog.Warningf("failed to read Cat L3 tasks for resctrl group %s, err: %s", group, err)
+		}
+	}
+
+	taskIds := map[string][]int{}
+	podsMeta := r.resManager.statesInformer.GetAllPods()
+	for _, podMeta := range podsMeta {
+		pod := podMeta.Pod
+		// only Running and Pending pods are considered
+		if pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending {
+			continue
+		}
+
+		// only uniapiext-QoS-specified pod are considered
+		podQOSCfg := getPodResourceQOSByQOSClass(pod, qosStrategy, r.resManager.config)
+		if podQOSCfg.ResctrlQoS.Enable == nil || !(*podQOSCfg.ResctrlQoS.Enable) {
+			klog.V(5).Infof("pod %v with qos %v disabled resctrl", util.GetPodKey(pod), extension.GetPodQoSClass(pod))
+			continue
+		}
+		if group := getPodResctrlGroup(pod); group != UnknownResctrlGroup {
+			ids := getPodCgroupNewTaskIds(podMeta, curTaskMaps[group])
```

Comment:
add a TODO here，this will be much simpler when we use runtime hooks.

Reply:
Done